notebooks/working/_03-tb-deconfounder-predictive-checks.md

```diff
@@ -0,0 +1,683 @@
+            show=show, close=close)
+    return overall_p_value, sampled_pvals, obs_pvals
+
+```
+
+## Extract data for the factor model checks
+
+```python
+# Load the raw data
+df = pd.read_csv(DATA_PATH)
+```
+
+```python
+# Note the variables that take part in the drive alone utility
+# The following cell is taken from 5.0pmab-simulation-causal-graph.ipynb
+V_Drive_Alone =\
+    CausalGraphicalModel(
+        nodes=["Total Travel Distance",
+               "Total Travel Time",
+               "Total Travel Cost",
+               "Number of Autos",
+               "Number of Licensed Drivers",
+               "Utility (Drive Alone)"],
+         edges=[("Total Travel Distance","Total Travel Time"),
+                ("Total Travel Distance","Total Travel Cost"),
+                ("Total Travel Time", "Utility (Drive Alone)"), 
+                ("Total Travel Cost", "Utility (Drive Alone)"), 
+                ("Number of Autos", "Utility (Drive Alone)"),
+                ("Number of Licensed Drivers","Utility (Drive Alone)")
+    ]
+)
+
+# draw the causal model
+V_Drive_Alone.draw()
+```
+
+```python
+# Create a list of the variables in the drive alone utility
+drive_alone_variables =\
+    ['total_travel_distance',
+     'total_travel_cost',
+     'total_travel_time',
+     'num_cars',
+     'num_licensed_drivers'
+    ]
+
+# Create a sub-dataframe with those variables
+drive_alone_df =\
+    df.loc[df['mode_id'] == 1, drive_alone_variables]
+
+# Get the means and standard deviations of those variables
+drive_alone_means = drive_alone_df.mean()
+drive_alone_means.name = 'mean'
+
+drive_alone_stds = drive_alone_df.std()
+drive_alone_stds.name = 'std'
+
+# Look at the computed means and standard deviations
+print(pd.DataFrame([drive_alone_means, drive_alone_stds]))
+```
+
+## Specify the factor model that is to be checked
+
+In Wang and Blei's deconfounder technique, we fit a factor model to the variables in one's outcome model.
+
+The factor model being considered here is:
+
+$
+\begin{aligned}
+X_{\textrm{standardized}} &= Z * W + \epsilon\\
+\textrm{where } \epsilon &= \left[ \epsilon_1, \epsilon_2, ..., \epsilon_D \right]\\
+\epsilon_d &\in \mathbb{R}^{\textrm{N x 1}}\\
+\epsilon_d &\sim \mathcal{N} \left(0, \sigma \right) \forall d \in \left\lbrace 1, 2, ... D \right\rbrace\\
+Z &\in \mathbb{R}^{\textrm{N x 1}}\\
+Z &\sim \mathcal{N} \left(0, 1 \right)\\
+W &\in \mathbb{R}^{1 x D}\\
+W &\sim \mathcal{N} \left(0, 1 \right)\\
+N &= \textrm{Number of rows in X_standardized}\\
+D &= \textrm{Number of columns in X_standardized}
+\end{aligned}
+$
+
+```python
+# Note the number of dimensions
+num_dimensions = len(drive_alone_variables)
+
+# Specify the prior distributions for the factor
+# model of the standardized drive alone dataframe
+w_dist_prior = norm(loc=0, scale=1)
+z_dist_prior = norm(loc=0, scale=1)
+
+sigma_prior = 0.1
+epsilon_dists_prior =\
+    [norm(loc=0, scale=sigma_prior) for i in range(num_dimensions)]
```

Comment:
Any reason why list comprehension is needed here vs. `num_dimensions * norm(loc=0, scale=sigma_prior)`?
Maybe each element in the list resulting from list comprehension has a different memory address?

Reply:
To be honest, I have no idea what multiplying a scalar by a distribution would do.

One thing I think I could have equivalently done is to simply set `epsilon_dists_prior = norm(loc=0, scale=sigma_prior)` and then get the correct dimensions when I sampled from it by doing `epsilon_dists_prior.rvs(size=(num_drive_alone_obs, num_dimensions, NUM_PRIOR_SAMPLES))`.

The list comprehension was the first thing that came to my mind because I knew for a fact that it would work. I def think this list comprehension should be replaced if possible when refactoring.

Reply:
Sorry what I meant was `num_dimensions * [norm(loc=0,scale=sigma_prior)]`

Reply:
Oh, got ya. That would probably have the same effect!